### ISSUE TITLE
No intellisense available for razor

### ISSUE THREAD

**Issue by egamma** [dotnet/vscode-csharp#168]
_From @Felsig on April 8, 2016 23:51_
- VSCode Version: 0.10.11 (0.10.11)
- OS Version: OSX El Capitan 10.11.2 

Specifically when writing code like @Model.Id or the similar I get no sort of intellisense. I assumed that would be a feature as the website states support for Razor. Is this just not a feature?

_Copied from original issue: Microsoft/vscode#5113_

**Comment by DustinCampbell** [comment]
Unfortunately, there is no way to get IntelliSense for Razor in VS Code through OmniSharp today. The Razor parser does not run as it does in Visual Studio. We will look at this in the future though.

**Comment by Felsig** [comment]
Thanks!

**Comment by eamodio** [comment]
Is there anyway to help out with this? Personally this is my number 1 issue with using VS Code for aspnet core development -- since refactoring/usage/etc is no longer safe because usages within cshtml aren't seen -- and cshtml views aren't even compiled on build (not sure if there is a way to make that happen with the new core stuff like there used to be).

**Comment by DustinCampbell** [comment]
@david-driscoll was looking at it awhile back over in https://github.com/OmniSharp/omnisharp-roslyn. Maybe drop by [Slack](https://goo.gl/Ovnqr1) and chat about it?

**Comment by eamodio** [comment]
@DustinCampbell Thanks -- I'll hop over there an find out.

**Comment by rasmus91** [comment]
Have the same Issues (but from Ubuntu) really missing that Razor Intellisense, it'd really make all the difference in the world.

Will i be able to spot progress on this feature here?

**Comment by eimik** [comment]
Agree. Now that we have tabs, razor-intellisense is the "one big thing" missing from VS Code for asp.net core devs.

**Comment by david-driscoll** [comment]
We'll be adding support to OmniSharp eventually to support Razor, sadly I don't have a specific timeline.  This is mainly because it's not a very easy problem to solve, but it is something we want to get working.

**Comment by mchomer** [comment]
Some updates?
I also really miss it.

**Comment by winuxue** [comment]
I need it also

**Comment by DustinCampbell** [comment]
To be clear, this is a fairly big work item. It is currently in our "future" milestone because we don't yet have a good sense of when it will happen..

**Comment by eamodio** [comment]
@DustinCampbell is there a design/approach in mind for this? It would be great if there was a rough outline of the work that needed to be done and some general guidelines for an approach. This is something I would love to see happen and would also help make it so, but I'm honestly not sure where to even get started (I've only barely started touching OmniSharp and I haven't found any real documentation/overview docs for it). If this was something that could be put together, I would hope others in the community would try to step up and help (I would with whatever time I could find).

**Comment by david-driscoll** [comment]
I have a branch that was the bare minimum to wire in the "basic" Razor runtime.  It's available https://github.com/OmniSharp/omnisharp-roslyn/tree/razorhost.

When I say bare minimum, basically I was able to get diagnostics returned from the Razor host, and there were a lot of them because we I hadn't wired in any of the references (including things like mscorelib, lol).

We have yet to be determined if that is the best way to do it, but we're working on it.

The work that needs to be done is basically
* Determine the best way to wire the Razor project into an "Mvc" assembly.
  * Basically my thought was, just look for any projects that have a reference to `Microsoft.AspNetCore.Mvc.Razor` which they would then get a new "sister"/"sibling" project that understands the Razor syntax.  
  * We then keep this projects references in sync with the references of the related project.  With that in theory any type will be reachable by both projects, so compiling Razor should be a little better.
* Figure out how to detect tag helpers
  * Wire in tag helpers (hopefully)
* Build out services that understand razor
  * Diagnostics (partially done already)
  * Autocomplete / Intellisense
  * Highlighting
    * Tag Helpers
    * Syntax Tokens

There are other challenges on the VSCode / Atom side, with properly merging the C# / HTML syntax, so that they show the correct context at all times.  This, I believe, is where a lot of the big work comes in because it involves work into VSCode itself.

My hope is this week to triage some bugs with the Atom plugin (I've not had much time lately!), and then dig into Razor support, and try to tie it into the DotNet Project System, which while is on it's way out, will be a valid proof of concept.

**Comment by egamma** [comment]
@david-driscoll 

 > I believe, is where a lot of the big work comes in because it involves work into VSCode itself.

This is correct and now that we have extracted the HTML support into its own language server, we will start in October with the nested languages support. We will start with CSS Intellisense, which comes from the CSS language server, inside HTML.

**Comment by uxsoft** [comment]
What is the status of this?:)

**Comment by eByte23** [comment]
Once again, any status update?

**Comment by DustinCampbell** [comment]
There is no movement on this at the moment. To be honest, this isn't at the top of the priority list as I'm mostly working on getting support for .NET Core .csproj projects finalized.

**Comment by eByte23** [comment]
Okay, just the only thing that is stopping me using vscode only atm. As For .net core 1.0 apps no razor compilation so I can't get intellisense and cannot get and view errors until runtime.
So I'm falling back to vs2015 all the time just for razor.
If I get free time I might play with the branch list above^.
Thanks

**Comment by DustinCampbell** [comment]
Any help would be welcome! :smile:

**Comment by karabaja4** [comment]
In the meantime, there is an extension for vscode that implements intellisense for the Model object at least :)

https://marketplace.visualstudio.com/items?itemName=schneiderpat.aspnet-helper

**Comment by mrtristan** [comment]
any plans for this?

**Comment by idarous** [comment]
I'm waiting for this feature too.
It should also be possible to debug and refactor razor template files like we can do in visual studio.

**Comment by ivankarpey** [comment]
I think even being able to format markup properly will be a huge bonus. Since now markup formatting isn't working and it's really annoying to format html tags by hands.

**Comment by SoheilAlizade** [comment]
please, I want this.

**Comment by SudoCat** [comment]
I'm going to jump on this bandwagon. As a front end developer, Visual Studio is gross. It's a horror to write JS/CSS/HTML in, but I'm have little choice but to use it for templating. I want to stop juggling between VS for C#/Razor and Atom for everything else.

**Comment by MichaelTontchev** [comment]
Sometimes I need 4 separate solutions open at the same time. VS cries when I try to do this. As such, being able to fully use Razor inside VS Code would be phenomenal.

**Comment by raimonomiar** [comment]
leave razor we gonna use some js framework< no I wont

**Comment by matjazmav** [comment]
+1 I need this...

**Comment by Calx-Takeno** [comment]
almost one year and half past... I am doubting this feature was buried deeply in vscode backlog tomb.

**Comment by idarous** [comment]
To my point of view it's an important feature. Good cshtml support is the only reason why I still use visual studio 2017 instead of vscode.

**Comment by SeriaWei** [comment]
Yes this is very helpfull

**Comment by erashishmodi** [comment]
Press CTRL+SHIFT+P
enter ext install aspnet-helper

**Comment by SeriaWei** [comment]
@erashishmodi I have tried it. But not powerful enough.

**Comment by donniefitz2** [comment]
This is the one essential thing that VS Code lacks. With Razor syntax and formatting, I'll use VS Code exclusively. Hoping this happens soon so I can stop running Windows in a VM for this purpose.

**Comment by mchomer** [comment]
Would be nice to get an status update on this. 
What is the big problem to solve this?
Is the Atom-base not able to perform something like this?

**Comment by DustinCampbell** [comment]
From the C# side, the difficulty is not having a Razor parser and infrastructure for doing this work. In Visual Studio, the system that handles Razor and .aspx pages is very complex and has deep ties to the Visual Studio editor.

**Comment by eByte23** [comment]
Hi all,
I'm throwing together a very simple extension that to start with will just do razor parsing and give back errors using Microsoft.AspNetCore.Razor.Language.

Then I'll slowly try and get intellisense etc working.

**Comment by mikes-gh** [comment]
This is not really vscode-version specific.
Can someone adjust the title of this issue to remove the version.

**Comment by tsaltd** [comment]
bump this up on the queue - a critically needed feature

**Comment by sstamoulis** [comment]
An update to this would be appreciated!
Is implementation of this feature even considered at this point?

**Comment by matjazmav** [comment]
This is deal breaker for me, until razor intellisense is not supported vscode and your extension are basically useless for anyone working with MVC. I'm really looking forward to use vscode as my primary IDE.

**Comment by mika76** [comment]
It's a bit strange that this issue has been rolling over since April 2016. It's Feb 2018 now. Seems to me it's more of an issue of slipping priorities than something being "hard" to do...

Please push this priority up - it's the one thing that is missing for a nice asp.net development experience. A lot of time can be spent in views and having no help from the editor makes it very frustrating.

**Comment by redonwhite** [comment]
Visual Studio Code has no IntelliSense for Razor.
Please prioritize this!

**Comment by mikes-gh** [comment]
@shanselman  bemoaning lack of razor support in vscode in [Feb 27th Community Standup](https://youtu.be/N_QZtEZRyuM?list=PL1rZQsJPBU2StolNg0aqvQswETPcYnNKL&t=3594)

**Comment by glat** [comment]
+1

**Comment by welkie** [comment]
I've been wondering where the support was for Razor in VSCode since I started using it last year. I understand it's hard to do since it is so tightly coupled to VS right now, but I'd love to see this prioritized. Heck, I'd even throw a few bucks into an OmniSharp [OpenCollective](https://opencollective.com/) or Patreon etc to see it happen. :P

A "workaround" I could suggest here is getting familiar with an SPA JS framework and TypeScript. Then you'll have Intellisense as you build each of the MVC layers. TypeScript isn't that hard to learn if you have experience with both C# and JS, and the SPA frameworks have stabilized a lot recently. We can pretty much count on React, Vue, and Angular to stick around for a while with good support. If you use the CLI to make a "with React" app, using `dotnet new react`, you'll see how they officially recommend you lay out your project to do this.

Another idea I had, if my aim was to just have a view layer instead of a full blown SPA, would be to make individual React apps for each page/route. Instead of passing data with view model using model binding from a controller action to a Razor view, I'd fetch the view model in the route's React app from the controller action, using it as JSON API only.

**Comment by jtlowe** [comment]
Yeah it’s pretty disappointing not to have support for something that is, in my opinion, fundamental to asp.net. And they continue to roll out all sorts of great things for Node, React and other stuff. 

I’ve thought about going your route of the SPA and API calls myself to get away from razor. But then again, razor makes it easy to do a lot of great thing.

**Comment by mjsabby** [comment]
@egamma At the Visual Studio for Mac UserVoice there seems to be a change in status for Razor Intellisense from whatever it was to "Planned" (On Jan 26th). I'm assuming this means C# + Razor mixed together. How are they doing it? [1]

Short of some VSCode support (or is that in the Language Server?) couldn't we make a composite language service whose first job is to take the Razor CSHTML file + all the references from the project, etc. do a parse rundown and fill in dummy value where Razor syntax + C# is discovered and in parallel send the generated C# code to the C# language server. And also send the Razor minus C# file to the HTML language server?

Maybe I'm missing something fundamental, but to me it would seem like this is solvable. Perhaps we're reluctant because it's somewhat messy?

Open question I can see are:

Q:  Who should federate to the 3 independent languages involved (Razor, HTML, C#)
A:  While it may seem like VSCode should take charge here and treat as CSHTML as HTML and then federate to Razor and C#, but I would argue it would lead to problems down the road when Razor syntax evolves beyond what HTML would be (that would be bad, but why pigeonhole ourselves).

Q: If there is a "Razor" service how should it talk to HTML and C# language services?
A: Some sort of Language Server Registry? Razor LS on opening a CSHTML page must query VSCode for a handle to the C# and HTML Language service?

Q: Should we explore an approach where we don't change VSCode and LangServer?
A: This could also work and most likely will be the shortest time to market although it requires copypasta of the C# Language Server and HTML Language Server behind like a proxy service that "fakes" the request so they think they are looking at complete documents even though there is a mediator

I personally have no time to invest in any of these solutions, but I wanted to write this down here so we can get an answer on what would be a next step here.

[1] https://visualstudio.uservoice.com/forums/563332-visual-studio-for-mac/suggestions/17389744-razor-intellisense

**Comment by david-driscoll** [comment]
Update!

The Microsoft MVP summit was last week, and I got a chance to sit with the guys that work on the Razor Language and Tooling.  It was a great conversation, and will hopefully lead to some form of Razor support in the near future.

1) The new tooling is open source at https://github.com/aspnet/Razor
  * The goal will be to adapt this tooling to work with OmniSharp (more below)
2) VSCode does not currently support mixed languages (as far as I know).  This means we will have to pick a few compromises.
  * Will HTML attribute tooling work correctly?
    * We're not sure.
  * Will Syntax highlighting work as expected?
    * We're not sure
  * Will Tag Helper highlighting work as expected?
    * We're not sure.
3) The goal is to support Razor across all editors that OmniSharp supports, and this will extend into our LSP support.

Today I've started a fork and branch https://github.com/OmniSharp/Razor/tree/omnisharp.  Right now this is a proof of concept, we have a few different requirements that have to be met, so we'll will probably be forked from Razor while we work through them.
* The razor assemblies follow the Roslyn practice of `internal all the things`, we have to find the API's we require and expose them.
* The razor assemblies are strong named, because they are consumed by visual studio, and strong naming is contagious.  OmniSharp is not strong named, and probably won't be for the foreseeable future, as we consume assemblies that are not strong named.

**Comment by aeschli** [comment]
@david-driscoll I'm happy to help regarding the open points you mentioned regarding preserving the HTML tooling capabilities. There are various approaches to look at: Forwarding calls to the HTML service on the original document or on a virtual, preprocessed document or incorporating the HTML support...

**Comment by seriouz** [comment]
@david-driscoll When vscode has no support for mixed languages, how does the vue team gets vscode intellisense working for one-file vue templates with html, typescript and scss?

**Comment by verysimplenick** [comment]
any news?

**Comment by Gaulomatic** [comment]
I had to switch away from VSCode. From a user's standpoint, this feature is a no-brainer for being on top of the priority list. But sometimes users and developers do not agree on what is important and what's not :-)

**Comment by mikes-gh** [comment]
No commits since the initial one on the 7th March .  Its all gone quiet :disappointed:

**Comment by myrup** [comment]
I've switched away as well. Good news is razor intellisense is planned for Visual Studio For Mac.

**Comment by seriouz** [comment]
Why is this good news, when i am using vscode? Does VS Mac use OmniSharp aswell?

**Comment by esaramago** [comment]
I avoid to use Razor because of the lack of support in vscode xD

**Comment by myrup** [comment]
@seriouz it's good news to me that there will be an option for Mac users.

**Comment by mikes-gh** [comment]
If anyone is watching [Community Standup](https://live.asp.net/) this week lets fire this issue into the questions.  I know @shanselman is a big proponent of vscode and they did say they were going to do something about this issue in one of the recent standups.

**Comment by Gaulomatic** [comment]
Jetbrains Rider became a really good alternative for C# / ASP.NET (Core) development; especially when you are running macOS - which limits the choices by default. I found that OmniSharp can be really buggy in larger solutions, up to the point where it is hard to even work with. It's sad - but not the end of the world since we have alternatives.

**Comment by welkie** [comment]
@Gaulomatic Note that Rider uses Omnisharp as a back end for C# code completion.

**Comment by Gaulomatic** [comment]
@welkie Rider uses ReSharper as backend.

**Comment by david-driscoll** [comment]
@mikes-gh between preparing for speaking engagements, my two paying jobs and vacations it's been hard to find time to sit down and focus on it.  Believe me it's on my list of things to work on when I have some spare time.

**Comment by mikes-gh** [comment]
@david-driscoll  Sorry didn't realize there was no sponsorship from Microsoft on this.  The way they were speaking on the standup was they were in a position to push it along but if its all down to you that's not really fair.

**Comment by shanselman** [comment]
@damianedwards

**Comment by welkie** [comment]
@david-driscoll Don't be sorry. I know better than to take the work open source devs for granted and I'm sure many of us here do too. :)

**Comment by donniefitz2** [comment]
@david-driscoll Yeah, no need to be sorry. I mean, it'll be nice when it's there and it will be deeply appreciated.

**Comment by Andrew-MSFT** [comment]
I just want to chime in on the thread here that from the Visual Studio & .NET side we do provide sponsorship and resources for this project.  In the specific case of Razor business priorities caused us to add support to Visual Studio for Mac first (which is now in Preview).  As that is winding down, we're now starting conversations about Visual Studio Code support, but don't have any timelines or firm commitments we can make at this time.  As @david-driscoll indicated it's a non-trivial amount of work (as a point of reference the Visual Studio for Mac Razor editor has been under development for over 6 months and is still in Preview at this time)

**Comment by mitchlabrador** [comment]
@Andrew-MSFT Thanks for the update Andrew! Has your team been able to make any progress?

**Comment by Andrew-MSFT** [comment]
We've started to investigate, but unfortunately I don't have anything firmer than that to add at this time.

**Comment by Miclebrick** [comment]
Really want to use VS Code, and affording Jetbrains Rider would be tough, but it seems like the only option at this rate...

**Comment by Andrew-MSFT** [comment]
@Miclebrick have you looked to see if Visual Studio Community edition is an option?

**Comment by Miclebrick** [comment]
Sadly it's not, since I currently have Linux on my workstation

**Comment by myrup** [comment]
@roboriaan VS community works excellent for OS X as well :)

**Comment by seriouz** [comment]
@Andrew-MSFT So .cshtml files do have to support Intellisense for:
`HTML`, `C#`, `Razor`, `Javascript`, `Typescript?`

**Comment by Andrew-MSFT** [comment]
You should get support for HTML, JavaScript, CSS, and TypeScript in VS Code, it's just C# support that's missing at this time.

**Comment by mdx0101** [comment]
C# support would be highly appreciated.

Many thanks.

**Comment by gianlazz** [comment]
Wow I was shocked to find out that this wasn't supported. Super disappointed :/

**Comment by seriouz** [comment]
They showed us a internal work in progress on the community stand up. And it looked promising!

**Comment by mikes-gh** [comment]
They were very non-committal on the Stand up.  It is not a work in progress but an early investigation that may or may not turn into support for Razor in vs code.  They were very clear that they weren't making any promises.  I really hope that this proves to be achievable with the resources available and would appreciate an update here on the status of the investigation.

**Comment by mikes-gh** [comment]
Is there a github repo of the investigation?

**Comment by mika76** [comment]
Getting updated on the issue is so disappointing. ASP.NET is a priority of .NET, VS Code is a priority - but the single thing that is stopping bringing the two together is what ... a lack of priority? I mean this issue was opened in 2016! All we've gotten are a couple of teasers and then silence in between.

**Comment by devSoheilAlizadeh** [comment]
@mika76 They work hard to prepare this feature if you want to know the feature progress, join to the OmniSharp slack team. I asked about this feature in the twitter.
![image](https://user-images.githubusercontent.com/22220356/44396499-506ca980-a552-11e8-9e7f-155b6a301334.png)

I think OmniSharp isn't part of Microsoft and they are alone in the development of this extensions.

**Comment by SeriaWei** [comment]
That's why I'm still using Visual Studio.

**Comment by welkie** [comment]
@SeriaWei I'm more of a fan of Rider from JetBrains myself, if I want Razor support. I can't wait for VS Code to get it though. :pray:

**Comment by danroth27** [comment]
As @Andrew-MSFT mentioned previously we have started investigating VSCode support for Razor. We even showed a demo recently on the [ASP.NET Community Standup](https://youtu.be/7Eh_l7jEcCo?t=15m34s) of some of our early prototyping work. We've also reached out to various folks in the OmniSharp community to discuss how best to proceed with this effort. Nothing to share publicly yet, but we are working on it.

**Comment by ShadowDancer** [comment]
@danroth27 
Do You have anything like dev feed? :) I would be happy even with prototype demo.

**Comment by danroth27** [comment]
@ShadowDancer Not yet, but the intent is that the development work will be public and open source. We just have a few more hoops to jump through.

**Comment by gripusa** [comment]
sometimes, I lost the picture how Microsoft operates, they have a great product in the form of VSCode and they support many languages but they dont support a language which is developed by them and core to their development frameworks. If this is an issue with revenue cutting, surely , supporting Razor would not cut tot their budget because front-end developers would not be using Visual Studio any way. In short, it will be great to have Razor support in VSCode.

**Comment by SIkebe** [comment]
@DMW007 Have you seen this?
https://github.com/OmniSharp/omnisharp-vscode/pull/2554

**Comment by mika76** [comment]
@SIkebe wow this is great news - we're so close now 👍 👏 🎉

**Comment by Andrew-MSFT** [comment]
As you can see from the referenced Pull Request, we are working on adding Razor support in VS Code and will let you know as soon as there is a preview version available to try

**Comment by danroth27** [comment]
Razor tooling work for VS Code is now happening here: https://github.com/aspnet/Razor.VSCode. It will all get released as part of the standard C# extension, but if you're specifically interested in Razor support in VS Code then the [Razor.VSCode](https://github.com/aspnet/Razor.VSCode) repo is where the action is at (source code, issue tracking, etc). Enjoy!

**Comment by Lepike** [comment]
That is really good new @danroth27. Its going to be a game changer. 
Thanks for sharing!

/ Lars

**Comment by kokman** [comment]
Good job. Thanks

**Comment by danroth27** [comment]
The [C# extension for Visual Studio Code](https://marketplace.visualstudio.com/items?itemName=ms-vscode.csharp) released today (1.17.0) with preview support for Razor editing. Includes support for directive and C# completions. No support for tag helpers yet - that's next.

**Comment by welkie** [comment]
@danroth27 Y'all are amazing. I'm going to make it my duty to test this thing.

**Comment by ryhled** [comment]
This should include intellisense for stuff like "@RenderPage()" and so forth? Im fairly new to using vs code but tried both regular and insider with the c# extension (1.17.1) and the razorpage intellisense just gives me squigglies and 'Name renderbody does not exist in this context'. Works fine when building and launching. No issue in VS 2019 intellisense (if the project is opened there).

**Comment by ryancheung** [comment]
Tag helpers intellisense seems works now with the latest c# extension.

**Comment by usmanchaudary** [comment]
https://github.com/OmniSharp/omnisharp-vscode/issues/4575 

here i ran into a similar issue. 
and it worked for me

**Comment by SomeoneIsWorking** [comment]
Razor has nothing more than a simple syntax highlighting for me
Ubuntu 21.10
O# v1.24.0-beta1

**Comment by JoeRobich** [comment]
@SomeoneIsWorking I see you are using the beta release. Have you enabled "useModernNET" as well?

**Comment by SomeoneIsWorking** [comment]
> @SomeoneIsWorking I see you are using the beta release. Have you enabled "useModernNET" as well?

Yes. Also tried the new release, same result. 

```
<Project Sdk="Microsoft.NET.Sdk.Web">
  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
...
```

OmniSharp configuration is

    "omnisharp.path": "latest",
    "omnisharp.useModernNet": true,
    "omnisharp.enableRoslynAnalyzers": true,
    "omnisharp.disableMSBuildDiagnosticWarning": true,

General list of problems
- _ViewImports.cshtml is ignored (program works so the file is in correct place and format)
- Error view ignores file changes, squiggly lines remain at the same place after code edit and save
- Often times code analysis runs in a weird state: no errors displayed, putting cursor on some variables/properties show their info but not for others (No error in Output/OmniSharp Log when info display doesn't work). I think the items that don't show info are the ones imported from _ViewImports

**Comment by rd003** [comment]
Any progress on that. Razor intellisense is still not working.

**Comment by mikadumont** [comment]
If you're having issues with the IntelliSense experience please file a new issue.
